tests/src/test/spark330/scala/org/apache/spark/sql/rapids/utils/BackendTestSettings.scala

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*** spark-rapids-shim-json-lines
+{"spark": "330"}
+spark-rapids-shim-json-lines ***/
+package org.apache.spark.sql.rapids.utils
+
+import java.util
+
+import scala.collection.JavaConverters._
+import scala.reflect.ClassTag
+
+import com.nvidia.spark.rapids.TestStats
+
+import org.apache.spark.sql.rapids.utils.RapidsTestConstants.RAPIDS_TEST
+
+abstract class BackendTestSettings {
+
+  private val enabledSuites: java.util.Map[String, SuiteSettings] = new util.HashMap()
+
+  protected def enableSuite[T: ClassTag]: SuiteSettings = {
+    val suiteName = implicitly[ClassTag[T]].runtimeClass.getCanonicalName
+    if (enabledSuites.containsKey(suiteName)) {
+      throw new IllegalArgumentException("Duplicated suite name: " + suiteName)
+    }
+    val suiteSettings = new SuiteSettings
+    enabledSuites.put(suiteName, suiteSettings)
+    suiteSettings
+  }
+
+  private[utils] def shouldRun(suiteName: String, testName: String): Boolean = {
+    if (!enabledSuites.containsKey(suiteName)) {
+      return false
+    }
+
+    val suiteSettings = enabledSuites.get(suiteName)
+
+    val inclusion = suiteSettings.inclusion.asScala
+    val exclusion = suiteSettings.exclusion.asScala
+
+    if (inclusion.isEmpty && exclusion.isEmpty) {
+      // default to run all cases under this suite
+      return true
+    }
+
+    if (inclusion.nonEmpty && exclusion.nonEmpty) {
+      // error
+      throw new IllegalStateException(
+        s"Do not use include and exclude conditions on the same test case: $suiteName:$testName")
+    }
+
+    if (inclusion.nonEmpty) {
+      // include mode
+      val isIncluded = inclusion.exists(_.isIncluded(testName))
+      return isIncluded
+    }
+
+    if (exclusion.nonEmpty) {
+      // exclude mode
+      val isExcluded = exclusion.exists(_.isExcluded(testName))
+      return !isExcluded
+    }
+
+    throw new IllegalStateException("Unreachable code")
+  }
+
+  sealed trait ExcludeReason
+  case class UNKNOWN_ISSUE() extends ExcludeReason
```

Comment:
Is there a followup issue to investigate these?  -- Yes, check this: https://github.com/NVIDIA/spark-rapids/issues?q=is%3Aopen+is%3Aissue+author%3Abinmahone+label%3Abug . Gary and Sameer seem to have a plan to spread the investigating work. UNKNOWN_ISSUE is a strong sign of need investigating.

Reply:
I changed UNKNOWN_ISSUE to case object, however KNOWN_ISSUE and WONT_FIX_ISSUE is kept as case class because we need to initialize them with different issue link in the future.

Reply:
I still don't understand why we're not linking the issue URL.  That's why I said I don't see why we would ever want to use UNKNOWN_ISSUE -- it seems worthless, just a way to bypass the "mandatory" reason, and thus it's not really mandatory.  Why aren't we putting the link to these tracking issues in the reasons, just like we do for the integration tests when we skip/xfail them?

Reply:
To be clear, if we have tracking issues for tests that are failing, then we can use `KNOWN_ISSUE("https://github.com/NVIDIA/spark-rapids/issue/####")` for the reason. Multiple tests can point to the same issue, matching how the issues were filed.  If we don't have issues filed to track failing tests, and we don't know we're not fixing them, then there should be an issue filed and linked.  That's why I don't see a use-case for UNKNOWN_ISSUE.

Reply:
I get your point. As you can see, the current status is:

1. I tried first batch of suites, 92% of the test cases can pass so we'd like to enroll them in CI ASAP.
2. Still there're 8% test cases cannot pass. We need to exclude them to avoid breaking daily CI.
3. Personally I don't have the skills the identify the root cause for each ignored test case, so what I can do is to create an unbrella issue for each suite, like https://github.com/NVIDIA/spark-rapids/issues/10771, and await other experts to further investigate the root cause for each of the ignored test case. Notice that we may open up several different child issue for https://github.com/NVIDIA/spark-rapids/issues/10771, these child issues are the REAL REASON, not the umbrella issue.
4. Of course, I can mark all ignored test cased as KNOWN_ISSUE("https://github.com/NVIDIA/spark-rapids/issues/10771"), but this is another kind of problem because 10771 is not the actual root cause. In fact, this will it hard to differetiate between "This ignored test case requires investigating" and "This ignored test case is ignored with a known root cause, and it's awaiting fixing"

With above said, I think UNKNOWN_ISSUE is good sign to indicate a state where first-round investigation is required.

Reply:
We want a tracking issue in the code.  If you want to keep UNKNOWN_ISSUE vs. KNOWN_ISSUE I don't care much either way, but both must take a reason string which would almost always contain a Github issue URL to link to the tracking issue. This is what we do for integration tests when skipping/xfail, don't see why we would want to do something different here.

Reply:
It makes perfect sense to follow existing convention in integration test. I'll have modified the PR accordingly. Please check again